lib/runtime/wasmer/instance.go

```diff
@@ -196,60 +253,74 @@ func GetRuntimeVersion(code []byte) (version runtime.Version, err error) {
 	return version, nil
 }
 
-var (
-	ErrCodeEmpty      = errors.New("code is empty")
-	ErrWASMDecompress = errors.New("wasm decompression failed")
-)
+// Exec calls the given function with the given data
+func (in *Instance) Exec(function string, data []byte) (result []byte, err error) {
+	in.mutex.Lock()
+	defer in.mutex.Unlock()
 
-func setupVM(code []byte) (instance wasm.Instance,
-	allocator *runtime.FreeingBumpHeapAllocator, err error) {
-	if len(code) == 0 {
-		return instance, nil, ErrCodeEmpty
+	if in.isClosed {
+		return nil, ErrInstanceIsStopped
 	}
 
-	code, err = decompressWasm(code)
+	dataLength := uint32(len(data))
+	inputPtr, err := in.ctx.Allocator.Allocate(dataLength)
 	if err != nil {
-		// Note the sentinel error is wrapped here since the ztsd Go library
-		// does not return any exported sentinel errors.
-		return instance, nil, fmt.Errorf("%w: %s", ErrWASMDecompress, err)
+		return nil, fmt.Errorf("allocating input memory: %w", err)
 	}
 
-	imports, err := importsNodeRuntime()
+	defer in.ctx.Allocator.Clear()
+
+	// Store the data into memory
+	memory := in.ctx.Memory.Data()
+	copy(memory[inputPtr:inputPtr+dataLength], data)
+
+	runtimeFunc, err := in.vm.Exports.GetFunction(function)
 	if err != nil {
-		return instance, nil, fmt.Errorf("creating node runtime imports: %w", err)
+		return nil, fmt.Errorf("%w: %s", ErrExportFunctionNotFound, function)
 	}
 
-	// Provide importable memory for newer runtimes
-	// TODO: determine memory descriptor size that the runtime wants from the wasm.
-	// should be doable w/ wasmer 1.0.0. (#1268)
-	memory, err := wasm.NewMemory(23, 0)
+	castedInputPointer, err := safeCastInt32(inputPtr)
 	if err != nil {
-		return instance, nil, fmt.Errorf("creating web assembly memory: %w", err)
+		panic(err)
 	}
 
-	_, err = imports.AppendMemory("memory", memory)
+	castedDataLength, err := safeCastInt32(dataLength)
 	if err != nil {
-		return instance, nil, fmt.Errorf("appending memory to imports: %w", err)
+		panic(err)
```

Comment:
so is this [issue](https://github.com/ChainSafe/gossamer/issues/3216) supposed to change this line to return an error?

Reply:
yes this is what I had in mind. Then whoever calls exec can handle different situations with respect to specific errors. Or even at the very least give helpful output